dot/parachain/dispute/participation.go

```diff
@@ -0,0 +1,372 @@
+type Participation interface {
+	// Queue a dispute for the node to participate in
+	Queue(context overseer.Context, request ParticipationRequest, priority ParticipationPriority) error
+
+	// Clear clears a participation request. This is called when we have the dispute result.
+	Clear(candidateHash common.Hash) error
+
+	// ProcessActiveLeavesUpdate processes an active leaves update
+	ProcessActiveLeavesUpdate(update overseer.ActiveLeavesUpdate) error
+
+	// BumpPriority bumps the priority for the given receipts
+	BumpPriority(ctx overseer.Context, receipts []parachainTypes.CandidateReceipt) error
+}
+
+type block struct {
+	Number uint32
+	Hash   common.Hash
+}
+
+// ParticipationHandler handles dispute participation.
+type ParticipationHandler struct {
+	runningParticipation sync.Map
+	workers              atomic.Int32
+
+	queue       Queue
+	sender      overseer.Sender // TODO: revisit this once we have the overseer
+	recentBlock *block
+
+	runtime parachain.RuntimeInstance
+
+	//TODO: metrics
+}
+
+const MaxParallelParticipation = 3
+
+func (p *ParticipationHandler) Queue(ctx overseer.Context,
+	request ParticipationRequest,
+	priority ParticipationPriority,
+) error {
+	if _, ok := p.runningParticipation.Load(request.candidateHash); ok {
+		return nil
+	}
+
+	// if we already have a recent block, participate right away
+	if p.recentBlock != nil && p.numberOfWorkers() < MaxParallelParticipation {
+		p.forkParticipation(&request, p.recentBlock.Hash)
+		return nil
+	}
+
+	blockNumber, err := getBlockNumber(ctx.Sender, request.candidateReceipt)
+	if err != nil {
+		return fmt.Errorf("get block number: %w", err)
+	}
+
+	comparator, err := NewCandidateComparator(&blockNumber, request.candidateReceipt)
+	if err != nil {
+		return fmt.Errorf("create candidate comparator: %w", err)
+	}
+
+	if err := p.queue.Queue(comparator, &request, priority); err != nil {
+		return fmt.Errorf("queue ParticipationHandler request: %w", err)
+	}
+
+	return nil
+}
+
+func (p *ParticipationHandler) Clear(candidateHash common.Hash) error {
+	p.runningParticipation.Delete(candidateHash)
+	p.workers.Add(-1)
+
+	if p.recentBlock == nil {
+		panic("we never ever reset recentBlock to nil and we already received a result, so it must have been set before. qed")
+	}
+
+	p.dequeueUntilCapacity(p.recentBlock.Hash)
+	return nil
+}
+
+func (p *ParticipationHandler) ProcessActiveLeavesUpdate(update overseer.ActiveLeavesUpdate) error {
+	if update.Activated == nil {
+		return nil
+	}
+
+	if p.recentBlock == nil {
+		p.recentBlock = &block{
+			Number: update.Activated.Number,
+			Hash:   update.Activated.Hash,
+		}
+
+		p.dequeueUntilCapacity(update.Activated.Hash)
+	} else {
+		if update.Activated.Number > p.recentBlock.Number {
+			p.recentBlock.Number = update.Activated.Number
+			p.recentBlock.Hash = update.Activated.Hash
+		}
+	}
+
+	return nil
+}
+
+func (p *ParticipationHandler) BumpPriority(ctx overseer.Context, receipts []parachainTypes.CandidateReceipt) error {
```

Comment:
could you remove the return since all returns are nil?

Reply:
I've removed it!